### ISSUE TITLE
Automatic snapshot assignment for newly added Devices in a Device Group

### ISSUE THREAD

**Issue by MarianRaphael** [FlowFuse/flowfuse#3260]
## Current Behaviour
In the current DevOps pipeline setup, when a Device Group is targeted for deployment, only the devices that are part of the group at the time of pipeline execution receive the new target snapshot. Devices added to the group after the execution do not automatically receive this snapshot.

## User Story
As a: User managing a fleet of devices

I want: To have the ability to assign a device to a device group after a pipeline execution, and for the device to automatically receive the snapshot that was deployed to the group during that execution.

So that: I can ensure all devices in a group are synchronized with the latest snapshot without the need to re-execute the pipeline for each new device added to the group.

## Scope
- Any device added to the group after the pipeline execution should automatically receive the snapshot that was most recently deployed to that group.
- Display the current target snapshot in the Device Group overview

## License
Enterprise Tier (as Device Groups)

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate

**Comment by MarianRaphael** [comment]
Requested by https://app-eu1.hubspot.com/contacts/26586079/record/0-2/6568939245

**Comment by Steve-Mcl** [comment]
Fortunately, I foreseen this requirement and the table design is structured such that this is easy to track/implement.

Also, discussed with Nick and at the time, and we determined this would not be implemented in the initial release as it had not been scoped and the consequences not considered.

Personally, I believe this should be done via an option when adding a device to a group since you may NOT actually want this to occur immediately and without warning that "things" could start moving/happening.  

In short, given the above, some thought and additional scope (UX and possibly UI) details adding to this issue.


<br>

Note for future self: through table `PipelineStageDeviceGroups` field `targetSnapshotId` is updated upon pipeline deploy

**Comment by Steve-Mcl** [comment]
Additionally, consideration about removing the target snapshot from a device removed from a group.

that then leads to: https://github.com/FlowFuse/flowfuse/issues/3226

**Comment by MarianRaphael** [comment]
> this would not be implemented in the initial release as it had not been scoped and the consequences not considered.

Absolutely right this here is a follow-up

> In short, given the above, some thought and additional scope (UX and possibly UI) details may need adding to this issue.

I want to keep this here simple and clear. The only UI scope here is "Display the current target snapshot in the Device Group overview". The other UI/UX Overview aspects will be designed and discussed here: https://github.com/FlowFuse/flowfuse/issues/3142

**Comment by Steve-Mcl** [comment]
> > In short, given the above, some thought and additional scope (UX and possibly UI) details may need adding to this issue.
> 
> I want to keep this here simple and clear.

@MarianRaphael I believe it would be prudent to consider this:

> Personally, I believe this should be done via an option when adding a device to a group since you may NOT actually want this to occur immediately and without warning that "things" could start moving/happening.

Consider the scenario:

* You have x devices that you need to add to a device group BUT FIRST you need to update the source snapshot
* The current snapshot is NOT what needs to be deployed (because changes have not yet been made to accommodate the 10 new devices)
* You CANNOT/DO NOT want to deploy the new changes to the existing devices without the other devices being online/ready to receive the new snapshot.


By simply adding a checkbox `"[x] Auto Deploy current device group snapshot 'version 2.1'?"` at the time of adding the device to a group can make this problem go away!

**Comment by MarianRaphael** [comment]
I'll ask the customer if this is a relevant scenario / use case for them. Currently, with devices assigned to an instance, it doesn't seem to be one. But I'll clarify this. For now, it's the (reduced) scope as described in the description.

**Comment by Steve-Mcl** [comment]
@MarianRaphael did you ask and did we receive feedback?

**Comment by griemenschneider** [comment]
@MarianRaphael invited me to comment on this one; in the scenario described above I would start with Y devices deployed in group A with snapshot 1; I would then add my new X devices to group B with snapshot 2; then I would use DevOps pipelines to deploy snapshot 3 to both group A and B to rectify and merge and operational differences; and then finally I would move all devices from group B back in to group A. During all of this I would expect devices associated with a group to automatically acquire that group's active snapshot. When removing devices from groups I would also expect them to lose their snapshot.

If any enhancements were made upon this request it would be to add a UI and API option to move a device between a device group; rather than un-assign and re-assign; this would minimize downtime of the Node-RED because there wouldn't be a 'null snapshot' period during the migration

**Comment by Steve-Mcl** [comment]
@griemenschneider Thanks for the feedback.

I have a working implementation of the above.

Unfortunately, there is (IMO) an undesirable side-effect in what you describe.

When you try to consolidate all devices back into group A, coupled with the desire for devices to have "lose their snapshot" "when removing devices from groups", devices snapshots will be unnecessarily cleared and then restored when they are added back into group A.

This could probably be negated by an option to ask whether to update the target snapshot of the devices (as per my previous statement)

@MarianRaphael ^

**Comment by griemenschneider** [comment]
Discussed this further with my team and we feel that the user intuitively expects a device inside a group to automatically assume that group's snapshot. If the user does not want a device to have any snapshot they will unassign from the group. When moving between groups, we would expect either from the UI or the API the ability to assign the device to another group. When the device arrives in the new group we would expect it to automatically assume the snapshot of the new group. If the platform must first unassign, restart the device, and the re-assign to accomplish this migration that behavior is indeed undesirable but ultimately acceptable.

**Comment by Steve-Mcl** [comment]
Verified operations on staging env:

Steps:
1. Create 3 devices
1. Create a device group and add Device 2+3
1. Deploy pipeline snapshot from Instance to Device Group
1. following deployment, 
   1. add device 1 to group: witness device 1 get the latest pipeline snapshot
   1. remove device 3 from group: witness device 3 get reset to starter flows
